### ISSUE TITLE
xsd.parseFile can't find schema file

### ISSUE THREAD

**Issue by ahwitz** [tngan/samlify#130]
With a fresh install of 2.3.0, I get `ENOENT: no such file or directory, open './schemas/saml-schema-protocol-2.0.xsd'` whenever anything calls isValidXml (libsaml.ts:700).

I also think line 702 should have a premature `return` with that `reject` - if the file load fails, it still tries to valid off the non-existent schema.

**Comment by tngan** [comment]
@ahwitz Interesting, possibly there is some problem in the file path, would be better to use absolute path instead of relative one. WIP

**Comment by tngan** [comment]
@ahwitz Fixed and patched v2.3.1.

**Comment by domenvajevec** [comment]
I'm getting an ```I/O warning : failed to load external entity "schemas/saml-schema-assertion-2.0.xsd"
I/O warning : failed to load external entity "schemas/xmldsig-core-schema.xsd"```
I believe the ```
    <import namespace="urn:oasis:names:tc:SAML:2.0:assertion"
        schemaLocation="schemas/saml-schema-assertion-2.0.xsd"/>``` in `schemas/saml-schema-protocol-2.0.xsd` needs to be updated too

**Comment by domenvajevec** [comment]
A little more info...
![image](https://user-images.githubusercontent.com/2238175/31143370-fee7e75a-a831-11e7-889b-77342bfe3935.png)

**Comment by tngan** [comment]
@domenvajevec I will standardise the error output in v.2.4. release to make better debugging experience. (#126) During the meantime, you can directly input your raw xml string to do debugging through the following script in Node.js REPL. 

```bash
$ node # running in your terminal
> const saml = require('samlify');
> saml.SamlLib.isValidXml('<test></test>').then(console.log);
```

**Comment by tngan** [comment]
@ahwitz Issue is reopened. Would you like to provide a way to reproduce it ?

Apparently, the problem is located in the `schemaLocation` of imported schema files in `saml-schema-protocol-2.0.xsd`.

**Comment by ahwitz** [comment]
Using...
```
  "_from": "samlify@latest",
  "_id": "samlify@2.3.1",
```

The following crude reduction of the code I was using...

```
var samlify = require('samlify');
var express = require('express');
var bodyParser = require('body-parser');
var unirest = require('unirest');
var fs = require('fs');

const sp = samlify.ServiceProvider({
    encPrivateKey: fs.readFileSync(/* PEM location */, 'utf8'),
    metadata: fs.readFileSync(/* SP metadata location */, 'utf8')
});

var app = express();
app.post("/saml", bodyParser.urlencoded({extended: true}), (req, res) =>
{
    unirest.get(/* IdP metadata */).end(function (response)
    {
        const idp = samlify.IdentityProvider({
            isAssertionEncrypted: true,
            metadata: response.body,
        });
        sp.parseLoginResponse(idp, 'post', req).then(function (parseResult)
        {
            console.log("good");
        });
    });
});

var server = require('http').createServer(app).listen(9931);
```

...prints out...

```
I/O warning : failed to load external entity "schemas/saml-schema-assertion-2.0.xsd"
I/O warning : failed to load external entity "schemas/xmldsig-core-schema.xsd"
...
(node:54216) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Invalid XSD schema
```

I've started building my own SAML SP from scratch because of this, workflow issues re: #124, and a digest problem similar to #125, so a) this is my trying to reverse-engineer old code and b) I probably won't be of much use here on out.

**Comment by tngan** [comment]
@ahwitz For #125, after discussed with @hammadzz, it's because of known issue in `xml-crypto` for missing namespace in canonizing the signed info xml.

https://github.com/yaronn/xml-crypto/issues/137
https://github.com/yaronn/xml-crypto/issues/94

**Comment by tngan** [comment]
@ahwitz During the meantime, what if you change the `schemaLocation` of `saml-schema-protocol-2.0.xsd` inside the node_module/samlify/build/schemas, by a) removing the `schema/`, b) using absolute path saying `build/schemas`.

**Comment by ahwitz** [comment]
I didn't have the xml-crypto#94 problem (saw that earlier in #125), but xml-crypto#137 might help - thanks for the links! May look into #137 later, but the "failed to load" error here was preventing me from testing that.

**Comment by tngan** [comment]
@ahwitz Schema validation is necessary, and the tests are passed without problems. I will try to reproduce the issue by following the code you have mentioned above, thanks for your input.

**Comment by ahwitz** [comment]
Just saw your second comment and was trying to work through it, but I'm not sure what you're asking for.

Those two `I/O warning` lines above come from the schema imports within the assertion schema (lines 12 and 14 of `saml-schema-assertion-2.0.xml`).  The "Invalid XSD schema" is thrown from `libxml-xsd index.js:52` - the `source` variable shows up as what seems to be an empty document, but I'm not sure what that variable's supposed to look like.

I haven't looked into the tests, but it might have something to do with the path to those schemas when the text executes?

**Comment by tngan** [comment]
@ahwitz Update: the issue can be reproduced in my side.

**Comment by tngan** [comment]
@ahwitz By following the workaround in node-libxml-xsd (https://github.com/albanm/node-libxml-xsd/issues/11), the path problem can be resolved.

In libsaml.ts line 700-701

```js
process.chdir(path.resolve(__dirname, '../schemas'));
xsd.parseFile(path.resolve('saml-schema-protocol-2.0.xsd') ...
```

and remove `schemas` folder in the `schemaLocation` in `saml-schema-assertion-2.0.xml`  I will patch it tonight.

**Comment by tngan** [comment]
@ahwitz Patched with v2.3.2
